code/forms/gridfield/GridFieldSiteTreeAddNewButton.php

```diff
@@ -13,34 +13,38 @@
 class GridFieldSiteTreeAddNewButton extends GridFieldAddNewButton
 	implements GridField_ActionProvider {
 
-
+	/**
+	 * Determine the list of classnames and titles allowed for a given parent object
+	 *
+	 * @param SiteTree $parent
+	 * @return boolean
+	 */
 	public function getAllowedChildren(SiteTree $parent) {
+		if(!$parent->canAddChildren()) return array();
 		$allowedChildren = $parent->allowedChildren();
-		if(empty($allowedChildren)) return array();
-
 		$children = array();
 		foreach($allowedChildren as $class) {
 			if(Config::inst()->get($class, "show_in_sitetree") === false) {
-				$children[$class] = Injector::inst()->create($class)->i18n_singular_name();
+				$instance = singleton($class);
+				// Note: Second argument to SiteTree::canCreate is proposed API
+				if($instance->canCreate(null, array('Parent' => $parent))) {

```

Comment:
The info is in the link I attached earlier. http://docs.silverstripe.org/en/changelogs/3.1.11

These optional parameters work perfectly fine in older versions of the framework, so it's not a dependency. It just makes the method work better, and is automatically supported if you upgrade to 3.1.12 or above.

If you use it on a lower level, it'll simply default to the old behaviour, because PHP doesn't care if you append extra parameters to a method call.